core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1263,8 +1269,7 @@ class DAGScheduler(
           if (runningStages.contains(stage)) {
             try { // cancelTasks will fail if a SchedulerBackend does not implement killTask
               taskScheduler.cancelTasks(stageId, shouldInterruptThread)
-              stage.latestInfo.stageFailed(failureReason)
-              listenerBus.post(SparkListenerStageCompleted(stage.latestInfo))
+              markStageAsFinished(stage, Some(failureReason))
```

Comment:
There's a slight change in behavior here: in the old code, we never removed the stage from `runningStages` here even though we posted a StageCompletion listener event.  I think this was probably a bug.  In this updated code, `markStagesAsFinished` will remove the stage from `runningStages`.